File: clients/typescript/src/themis-client.ts
/**
 * ThemisDB Native TypeScript/JavaScript Client
 * Binary Wire Protocol v1 implementation
 * Performance: 5-10x faster than HTTP REST
 */

import * as net from 'net';
import * as crypto from 'crypto';
import { EventEmitter } from 'events';

// =============================================================================
// Constants
// =============================================================================

const WIRE_MAGIC = 0x544D4442; // "TMDB" in ASCII
const WIRE_VERSION = 0x01;
const HEADER_SIZE = 12;
const CHECKSUM_SIZE = 4;
const MAX_PAYLOAD_SIZE = 64 * 1024 * 1024; // 64MB

enum OpCode {
  HELLO = 0x01,
  HELLO_ACK = 0x02,
  AUTH_REQUEST = 0x03,
  AUTH_RESPONSE = 0x04,
  AUTH_SUCCESS = 0x05,
  AUTH_FAILURE = 0x06,

  GET = 0x10,
  PUT = 0x11,
  DELETE = 0x12,
  BATCH_GET = 0x13,
  BATCH_PUT = 0x14,

  QUERY_AQL = 0x20,
  QUERY_RESULT = 0x21,
  QUERY_CURSOR = 0x22,
  CURSOR_NEXT = 0x23,
  CURSOR_CLOSE = 0x24,

  TRANSACTION_BEGIN = 0x30,
  TRANSACTION_COMMIT = 0x31,
  TRANSACTION_ABORT = 0x32,

  VECTOR_SEARCH = 0x40,
  GRAPH_TRAVERSE = 0x41,

  GEO_QUERY = 0x50,
  TIMESERIES_QUERY = 0x51,

  BPMN_START_PROCESS = 0x60,
  BPMN_TASK_COMPLETE = 0x61,
  BPMN_QUERY_INSTANCE = 0x62,

  ERROR = 0xf0,
  OK = 0xf1,
  PING = 0xfe,
  CLOSE = 0xff,
}

enum MessageFlags {
  NONE = 0x0000,
  SKIP_CHECKSUM = 0x0001,
  COMPRESSED = 0x0002,
  ENCRYPTED = 0x0004,
}

// =============================================================================
// Interfaces
// =============================================================================

interface WireFrameHeader {
  magic: number;
  version: number;
  opcode: number;
  flags: number;
  payload_size: number;
  sequence: number;
}

interface QueryOptions {
  limit?: number;
  offset?: number;
  timeout?: number;
  collect_stats?: boolean;
}

interface VectorSearchOptions extends QueryOptions {
  top_k?: number;
  metric?: 'euclidean' | 'cosine' | 'manhattan';
  filter?: Record<string, any>;
}

// =============================================================================
// Exception Classes
// =============================================================================

class WireProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'WireProtocolError';
  }
}

class AuthenticationError extends WireProtocolError {
  constructor(message: string) {
    super(message);
    this.name = 'AuthenticationError';
  }
}

class ConnectionError extends WireProtocolError {
  constructor(message: string) {
    super(message);
    this.name = 'ConnectionError';
  }
}

class ThemisDBError extends WireProtocolError {
  errorCode: number;
  detail: string;

  constructor(errorCode: number, message: string, detail: string = '') {
    super(message);
    this.name = 'ThemisDBError';
    this.errorCode = errorCode;
    this.detail = detail;
  }
}

// =============================================================================
// Wire Protocol Handler
// =============================================================================

class WireFrame {
  header: WireFrameHeader;
  payload: Buffer;

  constructor(
    opcode: OpCode,
    payload: Buffer = Buffer.alloc(0),
    flags: MessageFlags = MessageFlags.NONE,
    sequence: number = 0
  ) {
    this.payload = payload;
    this.header = {
      magic: WIRE_MAGIC,
      version: WIRE_VERSION,
      opcode,
      flags,
      payload_size: payload.length,
      sequence,
    };
  }

  toBuffer(): Buffer {
    const headerBuf = Buffer.alloc(HEADER_SIZE);
    let offset = 0;

    // Magic (4 bytes, big-endian)
    headerBuf.writeUInt32BE(this.header.magic, offset);
    offset += 4;

    // Version (1 byte), OpCode (1 byte), Flags (2 bytes)
    headerBuf.writeUInt8(this.header.version, offset);
    offset += 1;
    headerBuf.writeUInt8(this.header.opcode, offset);
    offset += 1;
    headerBuf.writeUInt16BE(this.header.flags, offset);
    offset += 2;

    // Payload Size (4 bytes)
    headerBuf.writeUInt32BE(this.header.payload_size, offset);
    offset += 4;

    // Calculate CRC32 checksum
    const crc32 = this.calculateCRC32(Buffer.concat([headerBuf.slice(0, 12), this.payload]));
    const checksumBuf = Buffer.alloc(4);
    checksumBuf.writeUInt32BE(crc32, 0);

    return Buffer.concat([headerBuf, this.payload, checksumBuf]);
  }

  static fromBuffer(data: Buffer): { frame: WireFrame; bytesRead: number } {
    if (data.length < HEADER_SIZE) {
      throw new WireProtocolError('Incomplete frame header');
    }

    let offset = 0;
    const magic = data.readUInt32BE(offset);
    offset += 4;

    if (magic !== WIRE_MAGIC) {
      throw new WireProtocolError(`Invalid magic: 0x${magic.toString(16)}`);
    }

    const version = data.readUInt8(offset);
    offset += 1;
    const opcode = data.readUInt8(offset);
    offset += 1;
    const flags = data.readUInt16BE(offset);
    offset += 2;
    const payloadSize = data.readUInt32BE(offset);
    offset += 4;

    if (payloadSize > MAX_PAYLOAD_SIZE) {
      throw new WireProtocolError(`Payload too large: ${payloadSize}`);
    }

    const totalSize = HEADER_SIZE + payloadSize + CHECKSUM_SIZE;
    if (data.length < totalSize) {
      throw new WireProtocolError('Incomplete frame');
    }

    const payload = data.slice(HEADER_SIZE, HEADER_SIZE + payloadSize);
    const checksum = data.readUInt32BE(HEADER_SIZE + payloadSize);

    // Verify checksum
    if (!(flags & MessageFlags.SKIP_CHECKSUM)) {
      const headerBuf = data.slice(0, HEADER_SIZE);
      const calculatedCRC = WireFrame.calculateCRC32Static(
        Buffer.concat([headerBuf, payload])
      );
      if (calculatedCRC !== checksum) {
        throw new WireProtocolError('Checksum mismatch');
      }
    }

    const frame = new WireFrame(opcode, payload, flags);
    return { frame, bytesRead: totalSize };
  }

  private calculateCRC32(data: Buffer): number {
    return WireFrame.calculateCRC32Static(data);
  }

  private static calculateCRC32Static(data: Buffer): number {
    // Simple CRC32 implementation (in production use crc32 library)
    let crc = 0xffffffff;
    for (let i = 0; i < data.length; i++) {
      crc = crc ^ data[i];
      for (let j = 0; j < 8; j++) {
        crc = (crc >>> 1) ^ ((crc & 1) ? 0xedb88320 : 0);
      }
    }
    return (crc ^ 0xffffffff) >>> 0;
  }
}

// =============================================================================
// ThemisDB Native Client
// =============================================================================

export class ThemisDBClient extends EventEmitter {
  private socket: net.Socket | null = null;
  private host: string;
  private port: number;
  private username: string;
  private password: string;
  private authenticated: boolean = false;
  private sequence: number = 0;
  private pendingRequests: Map<
    number,
    { resolve: (value: any) => void; reject: (error: Error) => void; timeout: NodeJS.Timeout }
  > = new Map();
  private receiveBuffer: Buffer = Buffer.alloc(0);
  private connectionTimeout: number;

  constructor(
    host: string = 'localhost',
    port: number = 8766,
    username: string = 'admin',
    password: string = 'admin',
    connectionTimeout: number = 30000
  ) {
    super();
    this.host = host;
    this.port = port;
    this.username = username;
    this.password = password;
    this.connectionTimeout = connectionTimeout;
  }

  /**
   * Connect to ThemisDB server
   */
  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.socket = net.createConnection(this.port, this.host);

      const timer = setTimeout(() => {
        this.socket?.destroy();
        reject(new ConnectionError(`Connection timeout after ${this.connectionTimeout}ms`));
      }, this.connectionTimeout);

      this.socket.on('connect', async () => {
        clearTimeout(timer);
        try {
          await this.authenticate();
          resolve();
        } catch (error) {
          reject(error);
        }
      });

      this.socket.on('data', (data: Buffer) => {
        this.handleData(data);
      });

      this.socket.on('error', (error: Error) => {
        reject(new ConnectionError(`Connection error: ${error.message}`));
      });

      this.socket.on('close', () => {
        this.authenticated = false;
        this.emit('disconnected');
      });
    });
  }

  /**
   * Disconnect from server
   */
  disconnect(): void {
    if (this.socket) {
      this.socket.destroy();
      this.socket = null;
      this.authenticated = false;
    }
  }

  /**
   * Authenticate with server
   */
  private async authenticate(): Promise<void> {
    const frame = new WireFrame(OpCode.HELLO, Buffer.from('ThemisDB/1.0'));
    await this.sendFrame(frame);

    // Expect HELLO_ACK
    const response = await this.waitForResponse(OpCode.HELLO_ACK);
    if (!response) {
      throw new AuthenticationError('Server did not acknowledge HELLO');
    }

    // Send credentials
    const credentials = Buffer.from(`${this.username}:${this.password}`);
    const authFrame = new WireFrame(OpCode.AUTH_REQUEST, credentials);
    await this.sendFrame(authFrame);

    // Wait for AUTH_SUCCESS
    const authResponse = await this.waitForResponse(OpCode.AUTH_SUCCESS);
    if (!authResponse) {
      throw new AuthenticationError('Authentication failed');
    }

    this.authenticated = true;
  }

  /**
   * Get document by key
   */
  async get(key: string): Promise<any> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(JSON.stringify({ key }));
    const frame = new WireFrame(OpCode.GET, payload);
    const response = await this.sendAndWait(frame);

    return JSON.parse(response.payload.toString('utf-8'));
  }

  /**
   * Put document
   */
  async put(key: string, value: any): Promise<void> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(JSON.stringify({ key, value }));
    const frame = new WireFrame(OpCode.PUT, payload);
    await this.sendAndWait(frame);
  }

  /**
   * Delete document
   */
  async delete(key: string): Promise<void> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(JSON.stringify({ key }));
    const frame = new WireFrame(OpCode.DELETE, payload);
    await this.sendAndWait(frame);
  }

  /**
   * Execute AQL query
   */
  async query(aql: string, options: QueryOptions = {}): Promise<any[]> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(JSON.stringify({ aql, options }));
    const frame = new WireFrame(OpCode.QUERY_AQL, payload);
    const response = await this.sendAndWait(frame);

    return JSON.parse(response.payload.toString('utf-8'));
  }

  /**
   * Vector search
   */
  async vectorSearch(
    collection: string,
    vector: number[],
    options: VectorSearchOptions = {}
  ): Promise<any[]> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(
      JSON.stringify({
        collection,
        vector,
        options: { top_k: 10, metric: 'cosine', ...options },
      })
    );
    const frame = new WireFrame(OpCode.VECTOR_SEARCH, payload);
    const response = await this.sendAndWait(frame);

    return JSON.parse(response.payload.toString('utf-8'));
  }

  /**
   * Geospatial query
   */
  async geoQuery(
    collection: string,
    lat: number,
    lon: number,
    radius_km: number,
    options: QueryOptions = {}
  ): Promise<any[]> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(
      JSON.stringify({ collection, lat, lon, radius_km, options })
    );
    const frame = new WireFrame(OpCode.GEO_QUERY, payload);
    const response = await this.sendAndWait(frame);

    return JSON.parse(response.payload.toString('utf-8'));
  }

  /**
   * Time-series query
   */
  async timeseriesQuery(
    collection: string,
    start_time: string,
    end_time: string,
    options: QueryOptions = {}
  ): Promise<any[]> {
    if (!this.authenticated) throw new ConnectionError('Not connected');

    const payload = Buffer.from(
      JSON.stringify({ collection, start_time, end_time, options })
    );
    const frame = new WireFrame(OpCode.TIMESERIES_QUERY, payload);
    const response = await this.sendAndWait(frame);

    return JSON.parse(response.payload.toString('utf-8'));
  }

  /**
   * Send frame and wait for response
   */
  private async sendAndWait(frame: WireFrame): Promise<WireFrame> {
    this.sequence++;
    frame.header.sequence = this.sequence;

    const promise = new Promise<WireFrame>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pendingRequests.delete(this.sequence);
        reject(new WireProtocolError('Request timeout'));
      }, 5000);

      this.pendingRequests.set(this.sequence, { resolve, reject, timeout });
    });

    await this.sendFrame(frame);
    return promise;
  }

  /**
   * Send frame to server
   */
  private sendFrame(frame: WireFrame): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.socket) {
        reject(new ConnectionError('Socket not connected'));
        return;
      }

      this.socket.write(frame.toBuffer(), (error) => {
        if (error) reject(error);
        else resolve();
      });
    });
  }

  /**
   * Handle incoming data
   */
  private handleData(data: Buffer): void {
    this.receiveBuffer = Buffer.concat([this.receiveBuffer, data]);

    while (this.receiveBuffer.length > 0) {
      try {
        const { frame, bytesRead } = WireFrame.fromBuffer(this.receiveBuffer);
        this.receiveBuffer = this.receiveBuffer.slice(bytesRead);

        if (frame.header.sequence > 0 && this.pendingRequests.has(frame.header.sequence)) {
          const pending = this.pendingRequests.get(frame.header.sequence)!;
          clearTimeout(pending.timeout);
          this.pendingRequests.delete(frame.header.sequence);

          if (frame.header.opcode === OpCode.OK) {
            pending.resolve(frame);
          } else if (frame.header.opcode === OpCode.ERROR) {
            const error = JSON.parse(frame.payload.toString('utf-8'));
            pending.reject(
              new ThemisDBError(error.code, error.message, error.detail)
            );
          } else {
            pending.resolve(frame);
          }
        } else {
          this.emit('message', frame);
        }
      } catch (error) {
        if (error instanceof WireProtocolError && error.message.includes('Incomplete')) {
          break; // Wait for more data
        }
        this.emit('error', error);
        break;
      }
    }
  }

  /**
   * Wait for specific response opcode
   */
  private waitForResponse(opcode: OpCode): Promise<WireFrame | null> {
    return new Promise((resolve) => {
      const timeout = setTimeout(() => resolve(null), 1000);
      const listener = (frame: WireFrame) => {
        if (frame.header.opcode === opcode) {
          clearTimeout(timeout);
          this.removeListener('message', listener);
          resolve(frame);
        }
      };
      this.on('message', listener);
    });
  }
}

// =============================================================================
// Export
// =============================================================================

export {
  WireFrame,
  OpCode,
  MessageFlags,
  WireProtocolError,
  AuthenticationError,
  ConnectionError,
  ThemisDBError,
};

File: clients/rust/src/themis_client.rs
use std::collections::HashMap;
use std::time::Duration;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::sync::Mutex;
use serde_json::{json, JsonValue};
use std::sync::Arc;

// Wire Protocol v1 Constants
const WIRE_MAGIC: u32 = 0x544D4442; // "TMDB"
const WIRE_VERSION: u8 = 0x01;
const HEADER_SIZE: usize = 12;
const CHECKSUM_SIZE: usize = 4;
const MAX_PAYLOAD_SIZE: usize = 64 * 1024 * 1024; // 64MB

// OpCode definitions
pub mod opcode {
    pub const HELLO: u8 = 0x01;
    pub const HELLO_ACK: u8 = 0x02;
    pub const AUTH_REQUEST: u8 = 0x03;
    pub const AUTH_RESPONSE: u8 = 0x04;
    pub const AUTH_SUCCESS: u8 = 0x05;
    pub const AUTH_FAILURE: u8 = 0x06;

    pub const GET: u8 = 0x10;
    pub const PUT: u8 = 0x11;
    pub const DELETE: u8 = 0x12;
    pub const BATCH_GET: u8 = 0x13;
    pub const BATCH_PUT: u8 = 0x14;

    pub const QUERY_AQL: u8 = 0x20;
    pub const QUERY_RESULT: u8 = 0x21;
    pub const QUERY_CURSOR: u8 = 0x22;
    pub const CURSOR_NEXT: u8 = 0x23;
    pub const CURSOR_CLOSE: u8 = 0x24;

    pub const TRANSACTION_BEGIN: u8 = 0x30;
    pub const TRANSACTION_COMMIT: u8 = 0x31;
    pub const TRANSACTION_ABORT: u8 = 0x32;

    pub const VECTOR_SEARCH: u8 = 0x40;
    pub const GRAPH_TRAVERSE: u8 = 0x41;

    pub const GEO_QUERY: u8 = 0x50;
    pub const TIMESERIES_QUERY: u8 = 0x51;

    pub const BPMN_START_PROCESS: u8 = 0x60;
    pub const BPMN_TASK_COMPLETE: u8 = 0x61;
    pub const BPMN_QUERY_INSTANCE: u8 = 0x62;

    pub const ERROR: u8 = 0xF0;
    pub const OK: u8 = 0xF1;
    pub const PING: u8 = 0xFE;
    pub const CLOSE: u8 = 0xFF;
}

// Message Flags
pub mod flags {
    pub const NONE: u16 = 0x0000;
    pub const SKIP_CHECKSUM: u16 = 0x0001;
    pub const COMPRESSED: u16 = 0x0002;
    pub const ENCRYPTED: u16 = 0x0004;
}

// Error types
#[derive(Debug)]
pub enum ThemisDBError {
    ConnectionError(String),
    AuthenticationError(String),
    ProtocolError(String),
    TimeoutError,
    SerializationError(String),
}

impl std::fmt::Display for ThemisDBError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ThemisDBError::ConnectionError(msg) => write!(f, "Connection error: {}", msg),
            ThemisDBError::AuthenticationError(msg) => write!(f, "Authentication error: {}", msg),
            ThemisDBError::ProtocolError(msg) => write!(f, "Protocol error: {}", msg),
            ThemisDBError::TimeoutError => write!(f, "Request timeout"),
            ThemisDBError::SerializationError(msg) => write!(f, "Serialization error: {}", msg),
        }
    }
}

impl std::error::Error for ThemisDBError {}

pub type Result<T> = std::result::Result<T, ThemisDBError>;

// WireFrame structure
pub struct WireFrame {
    pub version: u8,
    pub opcode: u8,
    pub flags: u16,
    pub sequence: u32,
    pub payload: Vec<u8>,
}

impl WireFrame {
    pub fn new(opcode: u8, payload: Vec<u8>, flags: u16, sequence: u32) -> Self {
        WireFrame {
            version: WIRE_VERSION,
            opcode,
            flags,
            sequence,
            payload,
        }
    }

    pub fn to_bytes(&self) -> Result<Vec<u8>> {
        let mut bytes = Vec::new();

        // Write header
        bytes.extend_from_slice(&WIRE_MAGIC.to_be_bytes());
        bytes.push(self.version);
        bytes.push(self.opcode);
        bytes.extend_from_slice(&self.flags.to_be_bytes());
        bytes.extend_from_slice(&(self.payload.len() as u32).to_be_bytes());

        // Write payload
        bytes.extend_from_slice(&self.payload);

        // Calculate and write checksum
        if (self.flags & flags::SKIP_CHECKSUM) == 0 {
            let checksum = crc32fast::hash(&bytes);
            bytes.extend_from_slice(&checksum.to_be_bytes());
        }

        Ok(bytes)
    }

    pub fn from_bytes(data: &[u8], offset: usize) -> Result<(Self, usize)> {
        if data.len() - offset < HEADER_SIZE {
            return Err(ThemisDBError::ProtocolError("Incomplete frame".to_string()));
        }

        let mut pos = offset;

        // Read header
        let magic = u32::from_be_bytes([data[pos], data[pos + 1], data[pos + 2], data[pos + 3]]);
        pos += 4;

        if magic != WIRE_MAGIC {
            return Err(ThemisDBError::ProtocolError(format!("Invalid magic: 0x{:x}", magic)));
        }

        let version = data[pos];
        pos += 1;

        let opcode = data[pos];
        pos += 1;

        let flags = u16::from_be_bytes([data[pos], data[pos + 1]]);
        pos += 2;

        let payload_size = u32::from_be_bytes([data[pos], data[pos + 1], data[pos + 2], data[pos + 3]]);
        pos += 4;

        let payload_size = payload_size as usize;
        if payload_size > MAX_PAYLOAD_SIZE {
            return Err(ThemisDBError::ProtocolError(format!(
                "Payload too large: {}",
                payload_size
            )));
        }

        // Check if we have complete payload
        if data.len() < offset + HEADER_SIZE + payload_size {
            return Err(ThemisDBError::ProtocolError("Incomplete payload".to_string()));
        }

        // Read payload
        let payload = data[pos..pos + payload_size].to_vec();
        pos += payload_size;

        // Read and verify checksum
        if (flags & flags::SKIP_CHECKSUM) == 0 {
            if data.len() < pos + CHECKSUM_SIZE {
                return Err(ThemisDBError::ProtocolError("Missing checksum".to_string()));
            }

            let checksum = u32::from_be_bytes([data[pos], data[pos + 1], data[pos + 2], data[pos + 3]]);
            pos += CHECKSUM_SIZE;

            let header_and_payload = &data[offset..offset + HEADER_SIZE + payload_size];
            let calculated_checksum = crc32fast::hash(header_and_payload);

            if calculated_checksum != checksum {
                return Err(ThemisDBError::ProtocolError("Checksum mismatch".to_string()));
            }
        }

        Ok((
            WireFrame {
                version,
                opcode,
                flags,
                sequence: 0,
                payload,
            },
            pos,
        ))
    }
}

// ThemisDB Client
pub struct ThemisDBClient {
    host: String,
    port: u16,
    username: String,
    password: String,
    stream: Arc<Mutex<Option<TcpStream>>>,
    authenticated: Arc<Mutex<bool>>,
    sequence: Arc<Mutex<u32>>,
    pending_requests: Arc<Mutex<HashMap<u32, tokio::sync::oneshot::Sender<WireFrame>>>>,
}

impl ThemisDBClient {
    pub fn new(host: impl Into<String>, port: u16, username: impl Into<String>, password: impl Into<String>) -> Self {
        ThemisDBClient {
            host: host.into(),
            port,
            username: username.into(),
            password: password.into(),
            stream: Arc::new(Mutex::new(None)),
            authenticated: Arc::new(Mutex::new(false)),
            sequence: Arc::new(Mutex::new(0)),
            pending_requests: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn connect(&self) -> Result<()> {
        let addr = format!("{}:{}", self.host, self.port);
        let stream = TcpStream::connect(&addr)
            .await
            .map_err(|e| ThemisDBError::ConnectionError(e.to_string()))?;

        {
            let mut s = self.stream.lock().await;
            *s = Some(stream);
        }

        self.authenticate().await?;

        // Start receive loop
        let stream_clone = self.stream.clone();
        let pending_requests_clone = self.pending_requests.clone();
        tokio::spawn(async move {
            Self::receive_loop(stream_clone, pending_requests_clone).await;
        });

        Ok(())
    }

    pub async fn disconnect(&self) -> Result<()> {
        let mut stream = self.stream.lock().await;
        if let Some(mut s) = stream.take() {
            let _ = s.shutdown().await;
        }
        Ok(())
    }

    async fn authenticate(&self) -> Result<()> {
        // Send HELLO
        let hello_frame = WireFrame::new(
            opcode::HELLO,
            b"ThemisDB/1.0".to_vec(),
            flags::NONE,
            0,
        );
        self.send_frame(&hello_frame).await?;

        // Send AUTH_REQUEST
        let credentials = format!("{}:{}", self.username, self.password);
        let auth_frame = WireFrame::new(
            opcode::AUTH_REQUEST,
            credentials.into_bytes(),
            flags::NONE,
            0,
        );
        self.send_frame(&auth_frame).await?;

        *self.authenticated.lock().await = true;
        Ok(())
    }

    pub async fn get(&self, key: &str) -> Result<JsonValue> {
        let payload = json!({ "key": key });
        let frame = WireFrame::new(
            opcode::GET,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        let response = self.send_and_wait(&frame).await?;
        serde_json::from_slice(&response.payload)
            .map_err(|e| ThemisDBError::SerializationError(e.to_string()))
    }

    pub async fn put(&self, key: &str, value: JsonValue) -> Result<()> {
        let payload = json!({ "key": key, "value": value });
        let frame = WireFrame::new(
            opcode::PUT,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        self.send_and_wait(&frame).await?;
        Ok(())
    }

    pub async fn delete(&self, key: &str) -> Result<()> {
        let payload = json!({ "key": key });
        let frame = WireFrame::new(
            opcode::DELETE,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        self.send_and_wait(&frame).await?;
        Ok(())
    }

    pub async fn query(&self, aql: &str, options: Option<JsonValue>) -> Result<JsonValue> {
        let mut payload = json!({ "aql": aql });
        if let Some(opts) = options {
            payload["options"] = opts;
        }

        let frame = WireFrame::new(
            opcode::QUERY_AQL,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        let response = self.send_and_wait(&frame).await?;
        serde_json::from_slice(&response.payload)
            .map_err(|e| ThemisDBError::SerializationError(e.to_string()))
    }

    pub async fn vector_search(&self, collection: &str, vector: Vec<f64>, options: Option<JsonValue>) -> Result<JsonValue> {
        let mut opts = options.unwrap_or_else(|| json!({}));
        if !opts.get("top_k").is_some() {
            opts["top_k"] = json!(10);
        }
        if !opts.get("metric").is_some() {
            opts["metric"] = json!("cosine");
        }

        let payload = json!({
            "collection": collection,
            "vector": vector,
            "options": opts
        });

        let frame = WireFrame::new(
            opcode::VECTOR_SEARCH,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        let response = self.send_and_wait(&frame).await?;
        serde_json::from_slice(&response.payload)
            .map_err(|e| ThemisDBError::SerializationError(e.to_string()))
    }

    pub async fn geo_query(&self, collection: &str, lat: f64, lon: f64, radius_km: f64, options: Option<JsonValue>) -> Result<JsonValue> {
        let mut payload = json!({
            "collection": collection,
            "lat": lat,
            "lon": lon,
            "radius_km": radius_km
        });
        if let Some(opts) = options {
            payload["options"] = opts;
        }

        let frame = WireFrame::new(
            opcode::GEO_QUERY,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        let response = self.send_and_wait(&frame).await?;
        serde_json::from_slice(&response.payload)
            .map_err(|e| ThemisDBError::SerializationError(e.to_string()))
    }

    pub async fn timeseries_query(&self, collection: &str, start_time: &str, end_time: &str, options: Option<JsonValue>) -> Result<JsonValue> {
        let mut payload = json!({
            "collection": collection,
            "start_time": start_time,
            "end_time": end_time
        });
        if let Some(opts) = options {
            payload["options"] = opts;
        }

        let frame = WireFrame::new(
            opcode::TIMESERIES_QUERY,
            serde_json::to_vec(&payload)
                .map_err(|e| ThemisDBError::SerializationError(e.to_string()))?,
            flags::NONE,
            self.next_sequence().await,
        );

        let response = self.send_and_wait(&frame).await?;
        serde_json::from_slice(&response.payload)
            .map_err(|e| ThemisDBError::SerializationError(e.to_string()))
    }

    // Helper methods

    async fn next_sequence(&self) -> u32 {
        let mut seq = self.sequence.lock().await;
        *seq += 1;
        *seq
    }

    async fn send_frame(&self, frame: &WireFrame) -> Result<()> {
        let bytes = frame.to_bytes()?;
        let mut stream = self.stream.lock().await;
        if let Some(s) = stream.as_mut() {
            s.write_all(&bytes)
                .await
                .map_err(|e| ThemisDBError::ConnectionError(e.to_string()))?;
        }
        Ok(())
    }

    async fn send_and_wait(&self, frame: &WireFrame) -> Result<WireFrame> {
        let (tx, rx) = tokio::sync::oneshot::channel();

        {
            let mut pending = self.pending_requests.lock().await;
            pending.insert(frame.sequence, tx);
        }

        self.send_frame(frame).await?;

        tokio::time::timeout(Duration::from_secs(5), rx)
            .await
            .map_err(|_| ThemisDBError::TimeoutError)?
            .map_err(|_| ThemisDBError::ProtocolError("Response channel closed".to_string()))
    }

    async fn receive_loop(
        stream: Arc<Mutex<Option<TcpStream>>>,
        pending_requests: Arc<Mutex<HashMap<u32, tokio::sync::oneshot::Sender<WireFrame>>>>,
    ) {
        let mut buffer = vec![0u8; 1024 * 1024];
        let mut receive_buffer = Vec::new();

        loop {
            let mut stream_guard = stream.lock().await;
            if let Some(s) = stream_guard.as_mut() {
                match s.read(&mut buffer).await {
                    Ok(0) => break,
                    Ok(n) => {
                        drop(stream_guard);

                        receive_buffer.extend_from_slice(&buffer[..n]);

                        let mut offset = 0;
                        while offset < receive_buffer.len() {
                            match WireFrame::from_bytes(&receive_buffer, offset) {
                                Ok((frame, new_offset)) => {
                                    offset = new_offset;

                                    if frame.sequence > 0 {
                                        let mut pending = pending_requests.lock().await;
                                        if let Some(tx) = pending.remove(&frame.sequence) {
                                            let _ = tx.send(frame);
                                        }
                                    }
                                }
                                Err(_) => break,
                            }
                        }

                        receive_buffer = receive_buffer[offset..].to_vec();
                    }
                    Err(_) => break,
                }
            } else {
                break;
            }
        }
    }
}
